File: packages/firebase_ml_vision/android/src/main/java/io/flutter/plugins/firebasemlvision/ImageLabeler.java
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package io.flutter.plugins.firebasemlvision;

import com.google.firebase.ml.vision.FirebaseVision;
import com.google.firebase.ml.vision.label.FirebaseVisionCloudImageLabelerOptions;
import com.google.firebase.ml.vision.label.FirebaseVisionImageLabeler;
import com.google.firebase.ml.vision.label.FirebaseVisionOnDeviceImageLabelerOptions;
import java.util.Map;

class ImageLabeler extends AbstractImageLabeler {
  private final FirebaseVisionImageLabeler labeler;

  ImageLabeler(FirebaseVision vision, Map<String, Object> options) {
    final String modelType = (String) options.get("modelType");
    if (modelType.equals("onDevice")) {
      labeler = vision.getOnDeviceImageLabeler(parseOptions(options));
    } else if (modelType.equals("cloud")) {
      labeler = vision.getCloudImageLabeler(parseCloudOptions(options));
    } else {
      final String message = String.format("No model for type: %s", modelType);
      throw new IllegalArgumentException(message);
    }
  }

  private FirebaseVisionOnDeviceImageLabelerOptions parseOptions(Map<String, Object> optionsData) {
    float conf = (float) (double) optionsData.get("confidenceThreshold");
    return new FirebaseVisionOnDeviceImageLabelerOptions.Builder()
        .setConfidenceThreshold(conf)
        .build();
  }

  private FirebaseVisionCloudImageLabelerOptions parseCloudOptions(
      Map<String, Object> optionsData) {
    float conf = (float) (double) optionsData.get("confidenceThreshold");
    return new FirebaseVisionCloudImageLabelerOptions.Builder()
        .setConfidenceThreshold(conf)
        .build();
  }

  @Override
  FirebaseVisionImageLabeler getImageLabeler() {
    return labeler;
  }
}

File: packages/firebase_ml_vision/ios/Classes/ImageLabeler.m
// Copyright 2019 The Chromium Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#import "FLTFirebaseMlVisionPlugin.h"

@import MLKitImageLabelingCommon;
@import MLKitImageLabeling;

@interface ImageLabeler ()
@property MLKImageLabeler *labeler;
@end

@implementation ImageLabeler
- (instancetype)initWithOptions:(NSDictionary *)options {
  self = [super init];
  if (self) {
    _labeler = [MLKImageLabeler imageLabelerWithOptions:[ImageLabeler parseOptions:options]];
  }
  return self;
}

- (void)handleDetection:(MLKVisionImage *)image result:(FlutterResult)result {
  [_labeler processImage:image
              completion:^(NSArray<MLKImageLabel *> *_Nullable labels, NSError *_Nullable error) {
                if (error) {
                  [FLTFirebaseMlVisionPlugin handleError:error result:result];
                  return;
                } else if (!labels) {
                  result(@[]);
                }

                NSMutableArray *labelData = [NSMutableArray array];
                for (MLKImageLabel *label in labels) {
                  NSDictionary *data = @{
                    @"confidence" : @(label.confidence),
                    @"entityID" : @(label.index),
                    @"text" : label.text,
                  };
                  [labelData addObject:data];
                }

                result(labelData);
              }];
}

+ (MLKImageLabelerOptions *)parseOptions:(NSDictionary *)optionsData {
  NSNumber *conf = optionsData[@"confidenceThreshold"];

  MLKImageLabelerOptions *options = [MLKImageLabelerOptions new];
  options.confidenceThreshold = conf;

  return options;
}

@end
